### PR TITLE
Fix handling of cases where a duplicate item is added to a set and related cleanups

### PR DESCRIPTION
**Pull request by keszybz**

### CODE REVIEW

File: src/network/networkd-ndisc.c [hunk 1]
```diff
@@ -972,13 +970,11 @@ int config_parse_ndisc_black_listed_prefix(
                 if (!a)
                         return log_oom();
 
-                r = set_ensure_put(&network->ndisc_black_listed_prefix, &in6_addr_hash_ops, a);
+                r = set_ensure_consume(&network->ndisc_black_listed_prefix, &in6_addr_hash_ops, TAKE_PTR(a));
                 if (r < 0) {
-                        log_syntax(unit, LOG_ERR, filename, line, r,
-                                   "Failed to store NDISC black listed prefix '%s', ignoring assignment: %m", n);
+                        return log_oom();
                         continue;
```
Comment (qhill) [inline src/network/networkd-ndisc.c]: continue after return?

Comment (keszybz) [reply src/network/networkd-ndisc.c]: Ooops.

**Comment by keszybz** [comment]
Updated with a fix for the issue found by @qhill and two failures found by CI (with test case for one of them).

**Comment by keszybz** [comment]
set_ensure_consume() was busted. I was even failing under valgrind the test I added in the patch. Somehow I managed to miss that ;(

File: src/sysv-generator/sysv-generator.c [hunk 2]
```diff
@@ -828,9 +827,8 @@ static int set_dependencies_from_rcnd(const LookupPaths *lp, Hashmap *all_servic
         if (r < 0)
                 return r;
 
-        STRV_FOREACH(p, sysvrcnd_path) {
-                for (i = 0; i < ELEMENTSOF(rcnd_table); i ++) {
-
+        STRV_FOREACH(p, sysvrcnd_path)
+                for (unsigned i = 0; i < ELEMENTSOF(rcnd_table); i ++) {
```
Comment (poettering) [inline src/sysv-generator/sysv-generator.c:831]: should probably be size_t as ELEMENTOFS() reurns size_t

File: src/sysv-generator/sysv-generator.c [hunk 3]
```diff
@@ -892,9 +890,8 @@ static int set_dependencies_from_rcnd(const LookupPaths *lp, Hashmap *all_servic
                                 }
                         }
                 }
-        }
 
-        for (i = 0; i < ELEMENTSOF(rcnd_table); i ++)
+        for (unsigned i = 0; i < ELEMENTSOF(rcnd_table); i++)
```
Comment (poettering) [inline src/sysv-generator/sysv-generator.c:894]: ditto

File: src/sysv-generator/sysv-generator.c [hunk 4]
```diff
@@ -911,7 +908,7 @@ static int set_dependencies_from_rcnd(const LookupPaths *lp, Hashmap *all_servic
         r = 0;
 
 finish:
-        for (i = 0; i < ELEMENTSOF(rcnd_table); i++)
+        for (unsigned i = 0; i < ELEMENTSOF(rcnd_table); i++)
```
Comment (poettering) [inline src/sysv-generator/sysv-generator.c:911]: ditto

File: src/sysv-generator/sysv-generator.c (hunk 2 above)
Comment (keszybz) [reply src/sysv-generator/sysv-generator.c:831]: It's a static table with 4 elements, `u8` would suffice ;) I don't think it's worth spinning up the CI for this...

**Comment by keszybz** [comment]
bionic-amd64 timed out, but I don't quite see why. I'll assume this is unrelated.

**Pull request merged**